internal/ostreeclient/ostreeclient.go

```diff
@@ -91,12 +100,53 @@ func (c *Client) SetDefaultDeployment(index int) error {
 	return err
 }
 
-func (c *Client) GetDeployment(osname string) (string, error) {
+func (c *Client) GetDeployment(stateroot string) (string, error) {
 	args := []string{"admin", "status"}
 	if c.ibi {
-		args = append(args, "--sysroot", "/mnt")
+		args = append(args, "--sysroot", common.OstreeDeployPathPrefix)
+	}
+
+	output, err := c.executor.Execute("ostree", args...)
+	if err != nil {
+		return "", err
+	}
+
+	// Example output:
+	//   # ostree admin status
+	//   * rhcos 9455b99374197f10c453eb96f1b66cea884b3dc16ce4bc753bdb7263602bb722.0
+	//       origin: <unknown origin type>
+	//     rhcos_4.15.0_rc.1 8ef186bc6407db2180726e32354c394c189c6e9be2c17839b313cf1fed3d5391.0
+	//       origin: <unknown origin type>
+	for _, line := range strings.Split(output, "\n") {
+		fields := strings.Fields(line)
+		if len(fields) == 0 {
+			continue
+		}
+
+		if fields[0] == "*" {
+			// Pop off the *, which indicates the currently booted deployment
+			fields = fields[1:]
+		}
+
+		if len(fields) < 2 {
+			continue
+		}
+
+		if fields[0] == stateroot {
+			// Return the deployment for the first matching stateroot
+			return fields[1], nil
+		}
+	}
```

Comment:
This parser could be expressed with regex, e.g.

```go
func getDeploymentHash(output, deployment string) (string, error) {
	matches := regexp.MustCompile(fmt.Sprintf(`(?m)^[\s\*]+%s\s+([a-f0-9]{64}.\d+)\s*$`, regexp.QuoteMeta(deployment))).FindStringSubmatch(output)

	if len(matches) > 2 {
		return "", fmt.Errorf("deployment %s found multiple times", deployment)
	} else if len(matches) == 0 {
		return "", fmt.Errorf("deployment %s not found", deployment)
	}

	return matches[1], nil
}
```

I personally find it easier to follow (especially if pasted to tools like regex101.com)

(feel free to ignore if you disagree)

Reply:
Given that this would only be parsing 4 lines at most, in practice, the regex doesn't buy us much in efficiency. And while I have a very long history of enjoying using regex in my code, there's an equally long history of others complaining about the readability of said regex 😃  I do like the suggested regex, though, so I'll think a bit

Reply:
>the regex doesn't buy us much in efficiency

Yeah this was mostly about simplicity and removing lines of code / logic, not about efficiency. But its admittedly subjective in this case, as you said, some people hate regex